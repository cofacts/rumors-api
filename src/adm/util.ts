import pino from 'pino';
import { HTTPError } from 'fets';
import type { RouterPlugin } from 'fets';
import { jwtVerify, createRemoteJWKSet, JWTPayload } from 'jose';

/** HTTP request headers to include in audit log */
const HEADERS_TO_INCLUDE = new Set([
  'cf-access-authenticated-user-email',
  'cf-connecting-ip',
  'cf-ipcountry',
]);

/**
 * JWT token content to include in audit log.
 *
 * Ref: https://developers.cloudflare.com/cloudflare-one/identity/authorization-cookie/application-token/
 * */
const TOKEN_KEYS_TO_INCLUDE = new Set([
  'aud',
  'exp',
  'iat',
  'country',
  'email', // For identity-based auth
  'common_name', // For service token auth
]);

/**
 * Verifying Cloudflare Access token.
 *
 * Ref: https://developers.cloudflare.com/cloudflare-one/identity/authorization-cookie/validating-json/#javascript-example
 */
const TEAM_DOMAIN = process.env.CLOUDFLARE_ACCESS_TEAM_DOMAIN;
const CERTS_URL = `${TEAM_DOMAIN}/cdn-cgi/access/certs`;
const JWKS = createRemoteJWKSet(new URL(CERTS_URL));

const logger = pino();

/**
 * Log request, response and user information.
 *
 * Abort request if it's not coming from Cloudflare Access.
 *
 * Ref: https://the-guild.dev/openapi/fets/server/plugins
 */
export function useAuditLog(): RouterPlugin<any, any> {
  return {
    async onRequest({ request }) {
      const requestInfo = {
        id: request.headers.get('cf-ray'),
        url: request.url,
        headers: Object.fromEntries(
          [...request.headers.entries()].filter(([key]) =>
            HEADERS_TO_INCLUDE.has(key)
          )
        ),
      };

      const token = request.headers.get('cf-access-jwt-assertion');
      if (!token) {
        throw new HTTPError(
          403,
          'Unauthorized',
          {},
          { message: 'Missing Cloudflare access token' }
        );
      }

      let payload: JWTPayload;
      try {
        payload = (
          await jwtVerify(token, JWKS, {
            issuer: TEAM_DOMAIN,
          })
        ).payload;
      } catch (e) {
        logger.error(e, 'Cloudflare access token verification failed');
        throw new HTTPError(403, 'Unauthorized', {});
      }

      const shouldIncludeBody =
        'common_name' in payload /* Called via service tokens */ ||
        request.headers.get('content-type') ===
          'application/json'; /* Probably from Swagger UI */

      logger.info(
        {
          ...requestInfo,
          user: Object.fromEntries(
            Object.entries(payload).filter(([key]) =>
              TOKEN_KEYS_TO_INCLUDE.has(key)
            )
          ),
          req: shouldIncludeBody ? await request.json() : undefined,
        },
        'Req start'
      );
    },

    async onResponse({ request, response }) {
      const shouldIncludeBody =
        response.ok &&
        (response.headers.get('content-type') ?? '').startsWith(
          'application/json'
        );

      logger.info(
        {
          id: request.headers.get('cf-ray'),
          url: request.url,
          status: response.status,
          res: shouldIncludeBody ? await response.json() : undefined,
        },
        'Req end'
      );
    },
  };
}
