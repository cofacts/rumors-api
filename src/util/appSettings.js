import yaml from 'js-yaml';
import fs from 'fs';

export class AppSettingsError extends Error {
  constructor(msg) {
    super(`[AppSettings] ${msg}`);
  }
}

export function validateYaml(yamlData) {
  if (!(yamlData instanceof Array)) {
    throw new AppSettingsError('App settings should be an array');
  }

  const dbAppIds = new Set();
  const headerAppSecrets = new Set();
  yamlData.forEach((appSetting, idx) => {
    // Check dbAppId
    //
    if (!appSetting.dbAppId)
      throw new AppSettingsError(`App setting ${idx} does not have dbAppId`);
    if (dbAppIds.has(appSetting.dbAppId))
      throw new AppSettingsError(`App setting ${idx} has duplicated dbAppId`);
    dbAppIds.add(appSetting.dbAppId);

    // Check headerAppSecrets
    //
    if (!(appSetting.headerAppSecrets instanceof Array))
      throw new AppSettingsError(
        `App setting ${idx} does not have headerAppSecrets array`
      );
    appSetting.headerAppSecrets.forEach(secret => {
      if (typeof secret !== 'string')
        throw new AppSettingsError(
          `App ${appSetting.dbAppId} has invalid secret`
        );
      if (headerAppSecrets.has(secret))
        throw new AppSettingsError(
          `App ${appSetting.dbAppId} has duplicated secret`
        );
      headerAppSecrets.add(secret);
    });
  });
}

export function loadAppSettings(path = process.env.APP_SETTINGS_PATH) {
  if (!path) throw new AppSettingsError('APP_SETTINGS_PATH is not specified');
  try {
    const yamlData = yaml.load(fs.readFileSync(path));
    validateYaml(yamlData);
    return yamlData;
  } catch (e) {
    console.error(e);
    return [];
  }
}
