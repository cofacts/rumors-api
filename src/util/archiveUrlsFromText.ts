/** Extract URLs from text and send to Internet Archive Wayback Machine */

import urlRegex from 'url-regex';
import { removeFBCLIDIfExist } from './scrapUrls';

export default async function archiveUrlsFromText(text: string) {
  const originalUrls = text.match(urlRegex()) || [];
  if (originalUrls.length === 0) return [];

  // Normalize URLs before sending to cache or scrapper to increase cache hit
  //
  const normalizedUrls = removeFBCLIDIfExist(originalUrls);

  const results = await Promise.all(
    normalizedUrls.map(async (url) => {
      const formData = new FormData();
      formData.append('url', url);
      formData.append('capture_screenshot', '1');
      formData.append('skip_first_archive', '1');
      formData.append('delay_wb_availability', '1'); // Help reduce load on IA servers

      return (
        await fetch('https://web.archive.org/save', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            Authorization: `LOW ${process.env.INTERNET_ARCHIVE_S3_ACCESS_KEY}:${process.env.INTERNET_ARCHIVE_S3_SECRET_KEY}`,
          },
          body: formData,
        })
      ).json();
    })
  );

  console.info(`[archiveUrlsFromText] Archiving ${results.length} URLs`);
  results.forEach((result, i) => {
    if (result.job_id) {
      console.info(`[archiveUrlsFromText] [ ${result.url} ]: ${result.job_id}`);
    } else {
      console.error(
        `[archiveUrlsFromText] [ ${normalizedUrls[i]} ]: ${JSON.stringify(
          result
        )}`
      );
    }
  });
  return results;
}
