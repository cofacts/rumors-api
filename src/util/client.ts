import elasticsearch from '@elastic/elasticsearch';

export default new elasticsearch.Client({
  node: process.env.ELASTICSEARCH_URL,
});

type ProcessMetaArgs<T> = {
  _id: string;
  _source: T;
  found: boolean;
  _score: number;
  highlight: object; // FIXME: use ES type
  inner_hits: object; // FIXME: use ES type
  sort: string;
  fields: object;
};

// Processes {_id, _version, found, _source: {...}} to
// {id, ..._source}.
//
export function processMeta<T extends object>({
  _id: id,
  _source: source,

  found, // for mget queries
  _score, // for search queries

  highlight, // search result highlighting. Should be {<fieldName>: ['...']}

  inner_hits, // nested query search result highlighting.

  sort, // cursor when sorted

  fields, // scripted fields (if any)
}: ProcessMetaArgs<T>) {
  if (found || _score !== undefined) {
    return {
      id,
      ...source,
      _cursor: sort,
      _score,
      highlight,
      inner_hits,
      _fields: fields,
    };
  }
  return null; // not found
}
