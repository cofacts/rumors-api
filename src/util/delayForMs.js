function delayForMs(delayMs) {
  return new Promise((resolve) => {
    setTimeout(() => resolve(), delayMs);
  });
}

export default delayForMs;
