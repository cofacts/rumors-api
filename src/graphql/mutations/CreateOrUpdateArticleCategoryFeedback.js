import { GraphQLString, GraphQLNonNull } from 'graphql';

import { assertUser, getContentDefaultStatus } from 'util/user';
import FeedbackVote from 'graphql/models/FeedbackVote';
import ArticleCategory from 'graphql/models/ArticleCategory';
import DataLoaders from 'graphql/dataLoaders';

import client from 'util/client';

export function getArticleCategoryFeedbackId({
  articleId,
  categoryId,
  userId,
  appId,
}) {
  return `${articleId}__${categoryId}__${userId}__${appId}`;
}

/**
 * @returns {ArticleCategory} The updated article category
 */
export async function createOrUpdateArticleCategoryFeedback({
  articleId,
  categoryId,
  vote,
  comment,
  user,

  // Use new instance of dataLoader if not provided
  loaders = new DataLoaders(),
}) {
  const now = new Date().toISOString();

  // (articleId, categoryId, userId, appId) should be unique
  // but user can update
  const id = getArticleCategoryFeedbackId({
    articleId,
    categoryId,
    userId: user.id,
    appId: user.appId,
  });

  await client.update({
    index: 'articlecategoryfeedbacks',
    type: 'doc',
    id,
    body: {
      doc: {
        score: vote,
        comment: comment,
        updatedAt: now,
      },
      upsert: {
        articleId,
        categoryId,
        userId: user.id,
        appId: user.appId,
        score: vote,
        createdAt: now,
        updatedAt: now,
        comment: comment,
        status: getContentDefaultStatus(user),
      },
    },
    refresh: 'true', // We are searching for articlecategoryfeedbacks immediately
  });

  const feedbacks = await loaders.articleCategoryFeedbacksLoader.load({
    articleId,
    categoryId,
  });

  const [positiveFeedbackCount, negativeFeedbackCount] = feedbacks
    .filter(({ status }) => status === 'NORMAL')
    .reduce(
      (agg, { score }) => {
        if (score === 1) {
          agg[0] += 1;
        } else if (score === -1) {
          agg[1] += 1;
        }
        return agg;
      },
      [0, 0]
    );

  const { body: articleCategoryUpdateResult } = await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
            int idx = 0;
            int categoryCount = ctx._source.articleCategories.size();
            for(; idx < categoryCount; idx += 1) {
              HashMap articleCategory = ctx._source.articleCategories.get(idx);
              if( articleCategory.get('categoryId').equals(params.categoryId) ) {
                break;
              }
            }

            if( idx === categoryCount ) {
              ctx.op = 'none';
            } else {
              ctx._source.articleCategories.get(idx).put(
                'positiveFeedbackCount', params.positiveFeedbackCount);
              ctx._source.articleCategories.get(idx).put(
                'negativeFeedbackCount', params.negativeFeedbackCount);
            }
          `,
        params: {
          categoryId,
          positiveFeedbackCount,
          negativeFeedbackCount,
        },
      },
    },
    _source: true,
  });

  /* istanbul ignore if */
  if (articleCategoryUpdateResult.result !== 'updated') {
    throw new Error(
      `Cannot article ${articleId}'s feedback count for feedback ID = ${id}`
    );
  }

  return articleCategoryUpdateResult.get._source.articleCategories.find(
    (articleCategory) => articleCategory.categoryId === categoryId
  );
}

export default {
  description: 'Create or update a feedback on an article-category connection',
  type: ArticleCategory,
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    categoryId: { type: new GraphQLNonNull(GraphQLString) },
    vote: { type: new GraphQLNonNull(FeedbackVote) },
    comment: { type: GraphQLString },
  },

  async resolve(
    rootValue,
    { articleId, categoryId, vote, comment },
    { user, loaders }
  ) {
    assertUser(user);

    const updatedArticleCategory = await createOrUpdateArticleCategoryFeedback({
      articleId,
      categoryId,
      vote,
      comment,
      user,
      loaders,
    });

    /* istanbul ignore if */
    if (!updatedArticleCategory) {
      throw new Error(
        `Cannot get updated article category with article ID = ${articleId} and category ID = ${categoryId}`
      );
    }

    return { articleId, ...updatedArticleCategory };
  },
};
