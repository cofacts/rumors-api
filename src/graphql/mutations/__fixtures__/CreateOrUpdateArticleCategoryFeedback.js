import { getArticleCategoryFeedbackId } from '../CreateOrUpdateArticleCategoryFeedback';

function generateEntry(data) {
  return {
    [`/articlecategoryfeedbacks/doc/${getArticleCategoryFeedbackId(
      data
    )}`]: data,
  };
}

export default {
  '/articles/doc/article1': {
    articleCategories: [
      {
        categoryId: 'category1',
        positiveFeedbackCount: 11,
        negativeFeedbackCount: 0,
      },
    ],
  },
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser2',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser3',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser4',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser5',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser6',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser7',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser8',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser9',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser10',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
  ...generateEntry({
    score: 1,
    articleId: 'article1',
    categoryId: 'category1',
    appId: 'testClient',
    userId: 'testUser11',
    createdAt: '2017-01-01T00:00:00.000Z',
  }),
};
