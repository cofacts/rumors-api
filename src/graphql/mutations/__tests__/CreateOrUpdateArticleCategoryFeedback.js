import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures, resetFrom } from 'util/fixtures';
import client from 'util/client';
import { getArticleCategoryFeedbackId } from '../CreateOrUpdateArticleCategoryFeedback';
import MockDate from 'mockdate';
import fixtures from '../__fixtures__/CreateOrUpdateArticleCategoryFeedback';

describe('CreateOrUpdateArticleCategoryFeedback', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('creates feedback on given article category', async () => {
    MockDate.set(1485593157011);
    const userId = 'test';
    const appId = 'test';
    const articleId = 'article1';
    const categoryId = 'category1';
    const comment = 'comment1';

    const { data, errors } = await gql`
      mutation($articleId: String!, $categoryId: String!, $comment: String!) {
        CreateOrUpdateArticleCategoryFeedback(
          articleId: $articleId
          categoryId: $categoryId
          vote: UPVOTE
          comment: $comment
        ) {
          articleId
          categoryId
          feedbackCount
          positiveFeedbackCount
          negativeFeedbackCount
          ownVote
          feedbacks {
            vote
            comment
          }
        }
      }
    `(
      {
        articleId,
        categoryId,
        comment,
      },
      { userId, appId, comment }
    );
    MockDate.reset();

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot();

    const id = getArticleCategoryFeedbackId({
      articleId,
      categoryId,
      userId,
      appId,
    });

    const conn = await client.get({
      index: 'articlecategoryfeedbacks',
      type: 'doc',
      id,
    });
    expect(conn._source).toMatchSnapshot();

    const article = await client.get({
      index: 'articles',
      type: 'doc',
      id: articleId,
    });
    expect(article._source.articleCategories).toMatchSnapshot();

    // Cleanup
    await client.delete({
      index: 'articlecategoryfeedbacks',
      type: 'doc',
      id,
    });
    await resetFrom(fixtures, '/articles/doc/article1');
  });

  it('updates existing feedback', async () => {
    MockDate.set(1485593157011);
    const userId = 'testUser';
    const appId = 'testClient';
    const articleId = 'article1';
    const categoryId = 'category1';

    const { data, errors } = await gql`
      mutation($articleId: String!, $categoryId: String!) {
        CreateOrUpdateArticleCategoryFeedback(
          articleId: $articleId
          categoryId: $categoryId
          vote: DOWNVOTE
        ) {
          articleId
          categoryId
          feedbackCount
          positiveFeedbackCount
          negativeFeedbackCount
          ownVote
          feedbacks {
            vote
            comment
          }
        }
      }
    `(
      {
        articleId,
        categoryId,
      },
      { userId, appId }
    );
    MockDate.reset();

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot();

    const id = getArticleCategoryFeedbackId({
      articleId,
      categoryId,
      userId,
      appId,
    });

    expect(
      (await client.get({ index: 'articlecategoryfeedbacks', type: 'doc', id }))
        ._source
    ).toMatchSnapshot();

    // Cleanup
    await resetFrom(fixtures, `/articlecategoryfeedbacks/doc/${id}`);
  });

  afterAll(() => unloadFixtures(fixtures));
});
