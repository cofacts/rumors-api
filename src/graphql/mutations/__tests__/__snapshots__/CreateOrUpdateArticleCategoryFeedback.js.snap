// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOrUpdateArticleCategoryFeedback creates feedback on given article category 1`] = `
Object {
  "CreateOrUpdateArticleCategoryFeedback": Object {
    "articleId": "article1",
    "categoryId": "category1",
    "feedbackCount": 12,
    "feedbacks": Array [
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": "comment1",
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
    ],
    "negativeFeedbackCount": 0,
    "ownVote": "UPVOTE",
    "positiveFeedbackCount": 12,
  },
}
`;

exports[`CreateOrUpdateArticleCategoryFeedback creates feedback on given article category 2`] = `
Object {
  "appId": "test",
  "articleId": "article1",
  "categoryId": "category1",
  "comment": "comment1",
  "createdAt": "2017-01-28T08:45:57.011Z",
  "score": 1,
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;

exports[`CreateOrUpdateArticleCategoryFeedback creates feedback on given article category 3`] = `
Array [
  Object {
    "categoryId": "category1",
    "negativeFeedbackCount": 0,
    "positiveFeedbackCount": 12,
  },
]
`;

exports[`CreateOrUpdateArticleCategoryFeedback updates existing feedback 1`] = `
Object {
  "CreateOrUpdateArticleCategoryFeedback": Object {
    "articleId": "article1",
    "categoryId": "category1",
    "feedbackCount": 12,
    "feedbacks": Array [
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": "comment1",
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "UPVOTE",
      },
      Object {
        "comment": null,
        "vote": "DOWNVOTE",
      },
    ],
    "negativeFeedbackCount": 1,
    "ownVote": "DOWNVOTE",
    "positiveFeedbackCount": 11,
  },
}
`;

exports[`CreateOrUpdateArticleCategoryFeedback updates existing feedback 2`] = `
Object {
  "appId": "testClient",
  "articleId": "article1",
  "categoryId": "category1",
  "createdAt": "2017-01-01T00:00:00.000Z",
  "score": -1,
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "testUser",
}
`;
